### PR TITLE
Change os.sameFileContent

### PR DESCRIPTION
**Pull request by juancarlospaco**
- Improve `os.sameFileContent`.
- Changed buffer size from hardcoded to `bufferSize` argument.
- Added `checkSize` argument to `os.sameFileContent`, if `checkSize` is `true` then checks the file sizes **before reading the files**.

# Note

- Fixes a potential file descriptor leak in `sameFileContent` when `readBuffer` would raise.

If you have 2 files, one file is 9 gigabytes, another file is 10 gigabytes, they can not have the same contents, 
but reading them almost entirely is slower than just stat the file size.

### CODE REVIEW

File: lib/pure/os.nim [hunk 1]
```diff
@@ -3217,6 +3184,43 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
+  ## if the files have different file sizes they can not have the same contents,
+  ## `checkSize = true` may be faster, specially for very big files.
+  ##
+  ## See also:
+  ## * `sameFile proc <#sameFile,string,string>`_
+  var
+    a, b: File
+  if not open(a, path1): return false
+  if not open(b, path2):
+    close(a)
```
Comment (timotheecour) [inline lib/pure/os.nim:3202]: is close needed after open fails? I don't know, but please double check (bonus points, clarify the docs for `open` on this aspect, maybe in another PR)

File: lib/pure/os.nim [hunk 2]
```diff
@@ -3217,6 +3184,43 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
+    return false
+  if checkSize and getFileInfo(a).size != getFileInfo(b).size: return false
```
Comment (timotheecour) [inline lib/pure/os.nim]: leaks a file descriptor here! This is such a frequent error, I see it all the time.
this is why `defer` should be used more widely IMO; the only remaining thing to do before using it more widely IMO is checking their runtime cost refs https://forum.nim-lang.org/t/4022#44758

File: lib/pure/os.nim [hunk 3]
```diff
@@ -3217,6 +3184,46 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
+    close(a)
+    return false
+  if checkSize and getFileInfo(a).size != getFileInfo(b).size:
+    close(a)
+    close(b)
+    return false
+  var bufA = alloc(bufferSize)
+  var bufB = alloc(bufferSize)
+  while true:
+    var readA = readBuffer(a, bufA, bufferSize)
```
Comment (timotheecour) [inline lib/pure/os.nim]: this is still not correct wrt fd leaks! the only correct way is defer or try/finally; but try/finally adds 2 nested scopes, and defer adds 0 syntactically.

readBuffer can raise `IOError`

File: lib/pure/os.nim (hunk 2 above)
Comment (juancarlospaco) [reply lib/pure/os.nim]: I know, had the commit, was waiting the CI.

File: lib/pure/os.nim (hunk 3 above)
Comment (juancarlospaco) [reply lib/pure/os.nim]: That code is pre-existent, but makes sense.

File: lib/pure/os.nim [hunk 4]
```diff
@@ -3217,6 +3184,48 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
+  ##
+  ## If `checkSize` is `true` then checks the file sizes *before reading the files*,
+  ## if the files have different file sizes they can not have the same contents,
+  ## `checkSize = true` may be faster, specially for very big files.
+  ##
+  ## See also:
+  ## * `sameFile proc <#sameFile,string,string>`_
+  var
+    a, b: File
+  if not open(a, path1): return false
```
Comment (timotheecour) [inline lib/pure/os.nim]: * this should raise when you can't `open` (as in before this PR) instead of silently ignoring errors
* ditto with path2, taking care of correctness wrt fd leaks
* maybe wrapping everything into 2 nested try/finally would be the simplest?

File: lib/pure/os.nim [hunk 5]
```diff
@@ -3217,6 +3184,48 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
+    close(b)
+    return false
+  var bufA = alloc(bufferSize)
+  var bufB = alloc(bufferSize)
+  try:
+    while true:
+      var readA = readBuffer(a, bufA, bufferSize)
+      var readB = readBuffer(b, bufB, bufferSize)
+      if readA != readB:
+        result = false
```
Comment (timotheecour) [inline lib/pure/os.nim]: nit: can remove `result = false` since it's false by default

File: lib/pure/os.nim [hunk 6]
```diff
@@ -3217,6 +3184,48 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
+        break
+      if readA == 0:
+        result = true
+        break
+      result = equalMem(bufA, bufB, readA)
+      if not result: break
+      if readA != bufferSize: break # end of file
```
Comment (timotheecour) [inline lib/pure/os.nim]: I'm not at all convinced that this is correct; eg, what about EINTR? eg https://stackoverflow.com/questions/53245721/are-interrupt-signals-dispatched-during-fread-and-fwrite-library-calls; eg: https://www.geeksforgeeks.org/fread-function-in-c/

> Return Value: The function returns the number of objects read successfully. If an error occurs, the return value can be less than count.

IMO the only robust way is by checking whether you're at the EOF, via `feof` (or rather its nim cross platform wrapper)

* https://www.tutorialspoint.com/c_standard_library/c_function_feof.htm

File: changelog.md [hunk 7]
```diff
@@ -38,6 +38,10 @@
 - Added `asyncdispatch.activeDescriptors` that returns the number of currently
   active async event handles/file descriptors
 
+- Changed `os.sameFileContent` buffer size from hardcoded to `bufferSize` argument.
+- Added `checkSize` argument to `os.sameFileContent`,
```
Comment (timotheecour) [inline changelog.md]: nit: group those under 1 bullet point? (but keep both sentences)

File: lib/pure/os.nim [hunk 8]
```diff
@@ -3217,6 +3184,45 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
         raiseOSError(osLastError(), path)
     rawToFormalFileInfo(rawInfo, path, result)
 
+proc sameFileContent*(path1, path2: string; checkSize = false; bufferSize = 8192.Positive): bool {.
```
Comment (timotheecour) [inline lib/pure/os.nim]: test needed in tests/stdlib/tos.nim, where you can specify extreme values like `bufferSize=1` to check correctness there

File: lib/pure/os.nim [hunk 9]
```diff
@@ -3217,6 +3184,41 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
+  ##
+  ## See also:
+  ## * `sameFile proc <#sameFile,string,string>`_
+  var a, b: File
+  var mustRead = true
+  var bufA = alloc(bufferSize)
+  var bufB = alloc(bufferSize)
+  try:  # readBuffer or open may or may not raise IOError.
+    if not(open(a, path1)) or not(open(b, path2)): mustRead = false
+    if checkSize and getFileInfo(a).size != getFileInfo(b).size: mustRead = false
```
Comment (timotheecour) [inline lib/pure/os.nim:3218]: i think this should be moved up 1 line, no need to `open` if `getFileInfo(x).size` differs

Comment (juancarlospaco) [reply lib/pure/os.nim:3218]: `getFileInfo` is taking `File`.

File: lib/pure/os.nim (hunk 3 above)
Comment (timotheecour) [reply lib/pure/os.nim]: indeed, it's pre-existant; then either fix it in this PR or file an issue so we don't forget, or send another PR for the sake of separating bugfixes from features

File: lib/pure/os.nim (hunk 6 above)
Comment (juancarlospaco) [reply lib/pure/os.nim]: is this ok? https://github.com/nim-lang/Nim/pull/16008/commits/5b068cbc656730b7663f8595db9079a2229f0900

File: lib/pure/os.nim (hunk 8 above)
Comment (juancarlospaco) [reply lib/pure/os.nim]: I think `bufferSize` must be a power of two ideally (?).

File: lib/pure/os.nim (hunk 6 above)
Comment (timotheecour) [reply lib/pure/os.nim]: > is this ok?

there are 2 things that are wrong:
* first, you should only do the check on a short read:
```nim
if readA != bufferSize:
  let enda = io.endOfFile(a)
  let endb = io.endOfFile(b)
  if enda != endb: result = false
  if enda or endb: break
```

* second, `io.endOfFile` is wrong, just filed https://github.com/nim-lang/Nim/issues/16011; maybe let's fix this first so your PR can use `io.endOfFile`, otherwise you'll introduce a regression until endOfFile is fixed?

from man fread:

>  If an error occurs, or the end-of-file is reached, the return value is a short object count (or zero).

from man feof:

> The function feof() tests the end-of-file indicator for the stream pointed to by stream, returning non-zero if it is set.  The end-of-file indicator may be
     cleared by explicitly calling clearerr(), or as a side-effect of other operations, e.g. fseek().

File: lib/pure/os.nim (hunk 8 above)
Comment (timotheecour) [reply lib/pure/os.nim]: what I mean is it's easier to test corner cases with small values of `bufferSize` (whether 1 or 2)

> must be a power of two ideally

why? man fread shows no such restriction

a test is needed, this function is sufficiently complex to justify it

File: lib/pure/os.nim (hunk 3 above)
Comment (timotheecour) [reply lib/pure/os.nim]: oh, can you add a comment in the PR body mentioning you fixed this bug? eg:
```
this PR also fixes a potential file descriptor leak in `sameFileContent` when `readBuffer` would raise
```

File: lib/pure/os.nim [hunk 10]
```diff
@@ -3217,6 +3184,46 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
 
+proc sameFileContent*(path1, path2: string; checkSize = false; bufferSize = 8192.Positive): bool {.
+    rtl, extern: "nos$1", tags: [ReadIOEffect], noWeirdTarget.} =
+  ## Returns true if both pathname arguments refer to files with identical
+  ## binary content.
+  ##
+  ## If `checkSize` is `true` then checks the file sizes *before reading the files*,
+  ## if the files have different file sizes they can not have the same contents,
+  ## `checkSize = true` may be faster, specially for very big files.
+  ## `bufferSize` must be a power of two.
```
Comment (timotheecour) [inline lib/pure/os.nim]: why?
(ditto for `assert (bufferSize and (bufferSize - 1)) == 0`; why needed?)

File: lib/pure/os.nim [hunk 11]
```diff
@@ -3217,6 +3184,46 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
+  ## See also:
+  ## * `sameFile proc <#sameFile,string,string>`_
+  assert (bufferSize and (bufferSize - 1)) == 0
+  var a, b: File
+  var mustRead = true
+  var bufA = alloc(bufferSize)
+  var bufB = alloc(bufferSize)
+  try:  # readBuffer or open may or may not raise IOError.
+    if not(open(a, path1)) or not(open(b, path2)): mustRead = false
+    if checkSize and getFileInfo(a).size != getFileInfo(b).size: mustRead = false
```
Comment (timotheecour) [inline lib/pure/os.nim:3218]: `if mustRead and checkSize and ...`

File: tests/stdlib/tos.nim [hunk 12]
```diff
@@ -539,3 +539,15 @@ block: # normalizeExe
     doAssert "foo/../bar".dup(normalizeExe) == "foo/../bar"
   when defined(windows):
     doAssert "foo".dup(normalizeExe) == "foo"
+
+
+block: # sameFileContent
+  writeFile("test_sameFileContent0.txt", "nim")
```
Comment (timotheecour) [inline tests/stdlib/tos.nim]: let f0 = "test_sameFileContent0.txt" (or const)
let f1 = "test_sameFileContent1.txt"
then use f0,f1

File: lib/pure/os.nim [hunk 13]
```diff
@@ -3217,6 +3184,44 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
+  var bufB = alloc(bufferSize)
+  try:  # readBuffer or open may or may not raise IOError.
+    if not(open(a, path1)) or not(open(b, path2)): mustRead = false
+    if mustRead and checkSize and getFileInfo(a).size != getFileInfo(b).size: mustRead = false
+    if mustRead:
+      while true:
+        var readA = readBuffer(a, bufA, bufferSize)
+        var readB = readBuffer(b, bufB, bufferSize)
+        if readA != readB: break
+        if readA == 0:
```
Comment (timotheecour) [inline lib/pure/os.nim]: seems incorrect; 0 could indicate error instead of end of file:
man fread:
> If an error occurs, or the end-of-file is reached, the return value is a short object count (or zero).

instead how about:
```nim
        if readA > 0:
          result = equalMem(bufA, bufB, readA)
          if not result: break
        if readA != bufferSize:
          let enda = io.endOfFile(a)
          let endb = io.endOfFile(b)
          if enda != endb: break
          if enda:
            result = true
            break
```

File: tests/stdlib/tos.nim [hunk 14]
```diff
@@ -539,3 +539,17 @@ block: # normalizeExe
...
+
+block: # sameFileContent
+  const f0 = "test_sameFileContent0.txt"
+  const f1 = "test_sameFileContent1.txt"
+  writeFile(f0, "nim")
+  copyFile(f0, f1)
+  doAssert sameFileContent(f0, f1)
+  doAssert sameFileContent(f0, f1, checkSize = true)
+  doAssert sameFileContent(f0, f1, bufferSize = 1024)
+  writeFile("test_sameFileContent0.txt", "?")
```
Comment (timotheecour) [inline tests/stdlib/tos.nim]: f0

File: lib/pure/os.nim (hunk 6 above)
Comment (juancarlospaco) [reply lib/pure/os.nim]: So the original code was kinda buggy already...
:(

File: lib/pure/os.nim [hunk 15]
```diff
@@ -3217,6 +3184,52 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
+  ##
+  ## See also:
+  ## * `sameFile proc <#sameFile,string,string>`_
+  var a, b: File
+  var mustRead = true
+  var bufA = alloc(bufferSize)
+  var bufB = alloc(bufferSize)
+  try:  # readBuffer or open may or may not raise IOError.
+    if not open(a, path1):
+      raise newException(IOError, "Can not open file: $1" % [path1])
```
Comment (timotheecour) [inline lib/pure/os.nim]: I now realize that before PR no exception would be raised, which seems buggy, eg:
`sameFileContent("nonexistant", "nonexistant")` would silently return false.
I prefer API's that raise rather than silently allow invalid inputs but given that this was previous behavior (witness the patch in koch.nim), we should let others chime in here. One possibility is a `checkArgs` but it also adds a param.

File: lib/pure/os.nim [hunk 16]
```diff
@@ -3217,6 +3184,52 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
+      #mustRead = false
+    if not open(b, path2):
+      raise newException(IOError, "Can not open file: $1" % [path2])
+      #mustRead = false
+    if checkSize and getFileInfo(a).size != getFileInfo(b).size: mustRead = false
+    if mustRead:
+      while true:
+        var readA = readBuffer(a, bufA, bufferSize)
+        var readB = readBuffer(b, bufB, bufferSize)
+        if readA != readB: break
```
Comment (timotheecour) [inline lib/pure/os.nim:3223]: 1 final complication is short-reads, where you'd get `readA != readB` yet the files have same content, eg because of EINTR that occured for 1 of the `readBuffer`.

There are multiple correct ways to handle this, eg using: `let readN = min(readA, readB)` but it requires a few more changes in the code to be correct. Not terribly complicated though.

Note that this is a pre-existing issue so could be left as TODO for another PR (and this comment left un-resolved)

File: lib/pure/os.nim (hunk 15 above)
Comment (juancarlospaco) [reply lib/pure/os.nim]: I am copying the solution used on `os` "checkDir".

File: koch.nim [hunk 17]
```diff
@@ -326,9 +326,15 @@ proc boot(args: string) =
     exec "$# jsonscript --nimcache:$# $# compiler" / "nim.nim" %
       [nimi, smartNimcache, args]
 
+    # https://github.com/nim-lang/Nim/pull/16008#discussion_r524961618
```
Comment (Araq) [inline koch.nim]: In the best case, this is completely unrelated code. In the worst case it shows you changed ``sameFileContent``'s behaviour.

File: koch.nim [hunk 18]
```diff
@@ -327,8 +327,8 @@ proc boot(args: string) =
       [nimi, smartNimcache, args]
 
     if sameFileContent(output, i.thVersion):
+      echo "Executables are equal: SUCCESS!"
```
Comment (timotheecour) [inline koch.nim]: revert this (innocuous) diff, ie move this line below like before PR

File: lib/pure/os.nim [hunk 19]
```diff
@@ -3217,6 +3184,57 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
         raiseOSError(osLastError(), path)
     rawToFormalFileInfo(rawInfo, path, result)
 
+proc sameFileContent*(path1, path2: string; checkSize = false; checkFiles = false; bufferSize = 8192.Positive): bool {.
```
Comment (timotheecour) [inline lib/pure/os.nim]: maybe leave as unresolved TODO for future PR:

since this PR exposes `bufferSize`, maybe it's the time to ask this:
I'm fine with exposing `bufferSize`, but can't we use the OS-specific optimal `bufferSize` when user doesn't explicitly set a value, instead of hardcoding it to `8192` ?
(yes, on some systems, it'd be 8192, but not all, and you get more portable/optimial code if user doesnt' need to specify it).

this could be:
```nim
proc getDefaultBufferSize*(): int = ...
proc sameFileContent*(path1, path2: string; checkSize = false; checkFiles = false; bufferSize = getDefaultBufferSize()): bool ...
```

Comment (juancarlospaco) [reply lib/pure/os.nim]: On modern times I dont know if theres such thing, for example you can add gigabytes of ram to a virtualbox with a click. :shrug: 

Do you know a reliable algo to do it?.

Still better than hardcoding everything.

Comment (timotheecour) [reply lib/pure/os.nim]: https://github.com/nim-lang/RFCs/issues/286

Comment (timotheecour) [reply lib/pure/os.nim]: => https://github.com/nim-lang/Nim/pull/16023

File: lib/pure/os.nim [hunk 20]
```diff
@@ -3187,38 +3187,56 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
+    rtl, extern: "nos$1", tags: [ReadIOEffect], noWeirdTarget.} =
+  ## Returns `true` if both pathname arguments refer to files with identical binary content.
+  ##
+  ## If `checkSize` is `true` then checks the file sizes *before reading the files*,
+  ## if the files have different file sizes they can not have the same contents,
+  ## `checkSize = true` may be faster, specially for very big files.
+  ## This does not fail if the file never existed in the first place, unless `checkFiles = true`.
+  ##
+  ## .. code-block:: nim
+  ##   echo sameFileContent("file0.txt", "file1.txt", checkSize = true, bufferSize = 999)
```
Comment (timotheecour) [inline lib/pure/os.nim]: turn into:
```nim
runnableExamples:
  doAssert sameFileContent(currentSourcePath, currentSourcePath, checkSize = true, bufferSize = 4096)
```

File: lib/pure/os.nim [hunk 21]
```diff
@@ -3187,38 +3187,56 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
-  if not open(b, path2):
+  var a, b: File
+  var mustRead = true
+  var bufA = alloc(bufferSize)
+  var bufB = alloc(bufferSize)
+  try:  # readBuffer or open may or may not raise IOError.
+    if not open(a, path1):
+      if checkFiles:
+        raise newException(IOError, "Can not open file: $1" % [path1])
+      else:
```
Comment (timotheecour) [inline lib/pure/os.nim]: * shorter:
```nim
if checkFiles:
  raise newException(IOError, "Can not open file: $1" % [path1])
mustRead = false
```
ditto elsewhere
* `[path]` => `path1`

File: lib/pure/os.nim [hunk 22]
```diff
@@ -3187,38 +3187,56 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
+        mustRead = false
+    if not open(b, path2):
+      if checkFiles:
+        raise newException(IOError, "Can not open file: $1" % [path2])
+      else:
+        mustRead = false
+    if checkSize and getFileInfo(a).size != getFileInfo(b).size: mustRead = false
+    if mustRead:
+      while true:
+        var readA = readBuffer(a, bufA, bufferSize)
```
Comment (timotheecour) [inline lib/pure/os.nim]: s/var/let/

File: lib/pure/os.nim [hunk 23]
```diff
@@ -3187,38 +3187,56 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
         raiseOSError(osLastError(), path)
     rawToFormalFileInfo(rawInfo, path, result)
 
-proc sameFileContent*(path1, path2: string): bool {.rtl, extern: "nos$1",
-  tags: [ReadIOEffect], noWeirdTarget.} =
-  ## Returns true if both pathname arguments refer to files with identical
-  ## binary content.
+proc sameFileContent*(path1, path2: string; checkSize = false; checkFiles = false; bufferSize = 8192.Positive): bool {.
```
Comment (timotheecour) [inline lib/pure/os.nim]: this cancels out what I did here https://github.com/nim-lang/Nim/pull/16023 ;instead:
```nim
proc sameFileContent*(path1, path2: string; checkSize = false; checkFiles = false; bufferSize = -1): bool =
  ## ... when bufferSize == -1, it will be inferred from getFileInfo(path1).blockSize`...
  var bufferSize = bufferSize
  ...
  if bufferSize == -1:
    bufSize = getFileInfo(a).blockSize
```

File: lib/pure/os.nim [hunk 24]
```diff
@@ -3187,38 +3187,54 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
   ## * `sameFile proc <#sameFile,string,string>`_
-  var
-    a, b: File
-  if not open(a, path1): return false
-  if not open(b, path2):
+  runnableExamples:
+    doAssert sameFileContent(currentSourcePath, currentSourcePath, checkSize = true, bufferSize = 4096)
+  var a, b: File
+  var mustRead = true
+  var bufferSize = if bufferSize == -1: getFileInfo(path1).blockSize else: bufferSize
```
Comment (timotheecour) [inline lib/pure/os.nim]: this would give `OSError` in case path1 is not stat-table, not honoring checkFiles.

the way it was written before you PR (from a File after it was already opened) avoids this, and avoids opening the file twice.

Comment (juancarlospaco) [reply lib/pure/os.nim]: The problem is that `alloc` only takes `Natural` must be outside the `try` block and needs `bufferSize` too.

Comment (timotheecour) [reply lib/pure/os.nim]: easy:
```nim
var bufA, bufB: pointer
try:
  ...
  let bufferSize = if bufferSize == -1: ... else: ...
  bufA = alloc(bufferSize)
  ...
```

File: lib/pure/os.nim [hunk 25]
```diff
@@ -3187,38 +3187,54 @@ proc getFileInfo*(path: string, followSymlink = true): FileInfo {.noWeirdTarget.
...
   ## * `sameFile proc <#sameFile,string,string>`_
-  var
-    a, b: File
-  if not open(a, path1): return false
-  if not open(b, path2):
+  runnableExamples:
+    doAssert sameFileContent(currentSourcePath, currentSourcePath, checkSize = true, bufferSize = 4096)
+  var a, b: File
+  var mustRead = true
+  var bufA = alloc(bufferSize)
```
Comment (timotheecour) [inline lib/pure/os.nim]: how can that work? that's a bug, you're calling `alloc` before updating bufferSize
the way i suggested in https://github.com/nim-lang/Nim/pull/16008#discussion_r528063361 works

File: lib/pure/os.nim (hunk 24 above)
Comment (juancarlospaco) [reply lib/pure/os.nim]: Yeah I know that but I get `Attempt to read from nil` sometimes as you see on a few tests.
